topics/performance-on-cpu.md

```diff
@@ -0,0 +1,222 @@
+
+#### Visualizing the recorded profile information using Flame Graphs
+
+[Flame graphs](http://www.brendangregg.com/flamegraphs.html) allow for a quick and accurate visualization of frequent code-paths. They can be generated using [Brendan Greg] open source programs on [github](https://github.com/brendangregg/FlameGraph), which create interactive SVGs from folded stack files.
+
+
+Specifically for perf we need to convert the generate perf.data into the captured stacks, and fold each of then into single lines.
+You can then render the on CPU flame graph with:
+
+    perf script > redis.perf.stacks
+    stackcollapse-perf.pl redis.perf.stacks > redis.folded.stacks
+    flamegraph.pl redis.folded.stacks > redis.svg
+
+By default perf script will generate a perf.data file in the current working directory. See the [perf script](https://linux.die.net/man/1/perf-script.html) documention for advanced usage.
+
+See [FlameGraph usage options](https://github.com/brendangregg/FlameGraph#options) for more advanced  Stack trace visualizations ( like the differential one ).
+
+
+#### Archiving and sharing recorded profile information
+
+So that analysis of perf.data contents can be possible on other machine than the one in which 
+collection happened, you need to export along with the perf.data file all object files 
+with build-ids found in the record data file. This can be easily done with the help of 
+[perf-archive.sh](https://github.com/torvalds/linux/blob/master/tools/perf/perf-archive.sh) script:
+
+    perf-archive.sh perf.data
+
+
+Now please run:
+
+    tar xvf perf.data.tar.bz2 -C ~/.debug
+
+wherever machine you need to run `perf report` on.
+
+
+### Sampling stack traces using bcc/BPF's profile
+    
+Similarly to perf, as of Linux kernel 4.9, BPF-optimized profiling is now fully available with the promise of lower overhead on CPU (as stack traces are frequency counted in
+kernel context) and disk I/O resources during profiling. 
+
+Apart from it, and relying solely on bcc/BPF's profile tool we've also removed the perf.data and intermediate steps if stack traces analysis is our main goal. You can use bcc's profile tool to output folded format directly, for flame graph generation:
+
+    $ /usr/share/bcc/tools/profile -F 999 -f --pid $(pgrep redis-server) --duration 60 > redis.folded.stacks
+
+In that manner, we've remove any preprocessing and can render the on CPU flame graph with a single command:
+
+    flamegraph.pl redis.folded.stacks > redis.svg
+
+
+### Visualizing the recorded profile information using Flame Graphs
+
+Call counts analysis with bcc/BPF
+------------
+
+A function may consume significant CPU cycles either
+because its code is slow or because it's frequently called. To answer at what rate are functions being called you can rely uppon call counts analysis using BCC's `funccount` tool:
```

Comment:
```suggestion
because its code is slow or because it's frequently called. To answer at what rate functions are being called, you can rely upon call counts analysis using BCC's `funccount` tool:
```